keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+- replace the full labels array of the first topology element
+`kudo update ... -p topology/0/labels=[ { key: "DCLabel", value: "dc1" }, { key: "Usage", value: "additionalItem" } ]`
+
+The `kudo update` command will not allow adding or removing a single entry from an array field, as this would make the command not idempotent anymore.
+
+We may decide to implement a `kudo patch` later on which is not idempotent an allows add/remove operations.
+
+#### Instance updates with a parameter file (`-P` option)
+
+When installing or updating an operator, users can use a file that contains parameter values: `kudo update --instance my-instance -P values.yaml`. We will merge the existing with the updated parameter using [JSON merge patch](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#use-a-json-merge-patch-to-update-a-deployment). This means, that to e.g. specifying a new array element, the whole array has to be provided. 
+
+#### Add Parameter Wizard `kudo package add parameter`
+
+The `add parameter` wizard needs to be rewritten and adjusted to the new format. In addition to the current data, it must ask the user for the data type of the new parameter and A JSON pointer in the existing structure. This, however, only applies to simple schemas as any advanced functionality like reusing existing schema definitions, schema combinations (`anyOf`, `oneOf` etc) are hard to express in the CLI.
+
+ It is unclear how practical generating a JSON-schema using a CLI wizard is. There is, however, a number of tools that can [generate a schema](https://www.liquid-technologies.com/online-json-to-schema-converter) given a valid JSON document.
+
+#### Parameter Listing `kudo package list parameters`
+
+The output for the parameter list must be adjusted. The first column of the `uitable` should contain an indented structure that allows a user to see the nested nature of the parameters.
+
+## Additional attributes in JSON-schema
+
+ - `trigger` Copied from params.yaml, describes the plan that is triggered
+ - `immutable` Copied from params.yaml, specifies if the field is updatable (Could this clash with a future addition to JSON-schema? Should we rename it?)
+ - `oldName` Provided for backwards compatibility. If provided, specifies the old name that was used in the flat parameters list in the previous version? Having the ability to map a parameter from the previous version to the current one, gives the operator developers the ability to update their schemas and rearrange/rename the parameters.
+
+### Trigger
+
+With the list of parameters, each parameter had a trigger attribute which specified a plan to trigger when the parameter was changed. With the nested structure, this gets more complicated as one "parameter branch" might have multiple different triggers defined which is clearly not something we want. Here is the new rule set:
+ - Every level of the structure *may* specify a trigger.
```

Comment:
why not only leaves to have a trigger? And no inheritance

Reply:
That would be an option, but at the same time it would 
- lead to a lot of repeated trigger attributes
- if we assume the default "When it's not set, it's `deploy`" rule, this could be easy to forget for params of other plans. I think it's easier to only specify the trigger on the `/Backup` object, and all parameters below that trigger the `backup` plan.

Only triggers on leaves is a valid strategy here, I think - Maybe inheritance is a premature optimization, i'm not fully sure. I still think i'd be nice to have it, from an operator developers perspective.

Reply:
yeah I am just afraid of some ambiguity in rich trees of params. But that said I don't even know how much is trigger used in the wild, if people usually leave it to default or ... ?

Reply:
That depends - It's mostly used on custom plans, i'm not sure how widely they are used. But I think custom plans is one of the USPs of KUDO, the ability to trigger a `backup` plan, or to execute a custom action is something that's not easily possibly with other solutions, and I think we should embrace it

Reply:
Yeah I actually don't feel strongly either way about this, it's definitely not a blocker for me

Reply:
> why not only leaves to have a trigger? And no inheritance

I tend to agree. While @ANeumann82 is right and it will _"lead to a lot of repeated trigger attributes"_ this is no different from what we have today. So, on the other hand, we save the operator developer some lines by defining it at the higher level, on the other, there is an increased mental tax to pay when figuring out the trigger inheritance. 

I think I'd rather save brain cycles rather than YAML lines. And we could always relax the rule and introduce inheritance later.